### PR TITLE
Ensure spn nodes are initialized

### PR DESCRIPTION
**Pull request by beckerg**
Signed-off-by: Greg Becker <gbecker@micron.com>

### CODE REVIEW

**Comment by beckerg** [comment]
This patch ensures spn nodes are always initialized, and now uses the "spn_managed" flag to to keep track of whether or not it's on csched's alist.  As a result, sp3_node_unlink_all() is now idempotent as there is at least one place (from prune trees) where it could be called many times for the same node.  We can also call it to reap a node from a join, and I added some code to check and ensure that node isn't also on the active dirty list and to deal with it if it is (we don't expect it to be on that list).

File: lib/cn/csched_sp3.h [hunk 1]
```diff
@@ -35,7 +35,7 @@ struct sp3_node {
     struct sp3_rbe   spn_rbe[wtype_MAX];
     struct list_head spn_rlink;
     struct list_head spn_alink;
-    bool             spn_initialized;
+    bool             spn_managed;
```
Comment (tristan957) [inline lib/cn/csched_sp3.h:38]: Any comment need updating?

File: lib/cn/csched_sp3.h [hunk 2]
```diff
@@ -53,6 +53,16 @@ struct sp3_tree {
     struct list_head spt_dtree_linkv[2];
 };
 
+static inline void
+sp3_node_init(struct sp3_node *spn)
+{
+    for (uint tx = 0; tx < NELEM(spn->spn_rbe); ++tx)
```
Comment (tristan957) [inline lib/cn/csched_sp3.h:59]: Should tx be a size_t? sizeof / sizeof = size_t?

**Review by tristan957 — COMMENTED** [review]
Make sure to include your comment in the final commit!

File: lib/cn/csched_sp3.h (hunk 2 above)
Comment (beckerg) [reply lib/cn/csched_sp3.h]: Done.

File: lib/cn/csched_sp3.h (hunk 1 above)
Comment (beckerg) [reply lib/cn/csched_sp3.h:38]: I could write a treatise, but the next guy should have to suffer like I did ;-)

**Review by tristan957 — APPROVED** [review]

**Review by nabeelmmd — APPROVED** [review]

**Pull request merged**